interactive/submit.py

```diff
@@ -0,0 +1,240 @@
+import os
+import shutil
+import subprocess
+import sys
+import tempfile
+from pathlib import Path
+
+from attrs import asdict
+from django.conf import settings
+from django.db import transaction
+
+from .models import AnalysisRequest
+from .opencodelists import _get_opencodelists_api
+
+
+# from jobserver.github import create_issue
+# from .emails import send_analysis_request_confirmation_email
+# from .slacks import notify_analysis_request_submitted
+
+
+@transaction.atomic()
+def submit_analysis(
+    *,
+    analysis,
+    backend,
+    creator,
+    project,
+    get_opencodelists_api=_get_opencodelists_api,
+    force=False,
+):
+    """
+    Create all the parts needed for an analysis
+
+    This will stay in job-server while create_commit() is intended to move to
+    an external service in the future.
+    """
+    # TODO: wrap form kwargs up in a dataclass?
+
+    # create an AnalysisRequest instance so we have a PK to use in various
+    # places, but we don't save it until we've written the commit and pushed
+    # it, so we can create the JobRequest this object needs
+    analysis_request = AnalysisRequest(
+        project=project,
+        created_by=creator,
+        title=analysis.title,
+        template_data=asdict(analysis),
+    )
+
+    # update the Analysis structure so we can pass around a single object
+    # if/when we pull the create_commit function out into another service
+    # this structure would be the JSON we send over
+    analysis.identifier = analysis_request.pk
+
+    sha, project_yaml = create_commit(
```

Comment:
So, we should probably think about the case where this commit/push succeeds, but the subsequent job request creation fails, and thus this *entire* transaction will not commit, and we've got a random commit unrelated to any existing AnalysisRequest or JobRequest in the git repo.

I don't think it's a big problem - the commit is never run - but we should probably be aware it's a possible failure mode.

Possibly, rather than wrapping the entire function, we could AnalysisRequest creation in its own transaction, and same for the JobRequest creation. That might leave us in a better state when something breaks.

But it might be better to leave as is, perhaps. Distributed systems are hard.

Reply:
Yeah, splitting this up is really tricky.  We don't actually save the AnalysisRequest until the very end so that we can rely on there being a related JobRequest, so we're just passing around an instance here, without any db row until the final `save()` call on it.